### DISCUSSION TITLE
Passing generics to `createTRPCProxyClient`?

### DISCUSSION THREAD

**Discussion by moltar** [trpc/trpc#3583] [General]
_Not sure if this is a bug, or if I am doing something wrong, so I am starting with a discussion rather than an issue._

---

I am having an issue with generics.

```ts
import { createTRPCProxyClient } from '@trpc/client';
import { AnyRouter } from '@trpc/server';

function factory<TRouter extends AnyRouter>() {
  return createTRPCProxyClient<TRouter>({
    links: [],
  });
}

export const trpc = factory();

```

Type error:

```
Argument of type '{ links: never[]; }' is not assignable to parameter of type 'CreateTRPCClientOptions<TRouter>'.
  Type '{ links: never[]; }' is not assignable to type 'CreateTRPCClientBaseOptions<TRouter>'.(2345)
```

Repro: https://stackblitz.com/edit/github-vsjln8?file=src/utils/trpc.ts

Any ideas what could be causing this?

Adding `@ts-ignore` makes the error go away, and the clients work as expected.

**Comment by KATT** [comment]
My gut feeling is that we do type-level checking that you set up your `TRPCClient`-options correctly with transformers if the server affects it so the types won't know at this stage if it requires a transformer or not.

This passes  for instance:

```ts
import { createTRPCProxyClient, httpBatchLink } from '@trpc/client';
import { AnyRouter } from '@trpc/server';

function factory<TRouter extends AnyRouter>() {
  return createTRPCProxyClient<TRouter>({
    links: [],
    transformer: null as any, // <---- hack
  });
}

export const trpc = factory();
```

https://stackblitz.com/edit/github-vsjln8-w1mxnv?file=src%2Futils%2Ftrpc.ts

This feature got added in #3289 by @QuiiBz & is nice because missing transformers is a very common pitfall when setting up, but I see that it's annoying here.

**Reply by KATT** [reply]
So I guess it's unexpected behavior that is caused by fixing another issue & I don't know if we can actually have both working.

**Reply by KATT** [reply]
Perhaps we should do some runtime checks instead of #3289 which can give you a better warning.

**Reply by moltar** [reply]
Nice one on the hack!

This also works:

```ts
transformer: undefined as never,
```

😁

**Reply by QuiiBz** [reply]
That’s strange because the transformer is optional, so TS shouldn’t complain about it. I’ll take a look at the issue.

**Reply by monarchwadia** [reply]
@KATT thanks for leaving your `factory` snippet above, it worked to solve my typing issues :-) you helped me avoid a circular dependency between my `commons` and `backend` modules.

i wanted to move my version of the `factory` into a `commons` library to keep things DRY. 

I had to keep it generic inside `commons` because the same code was also going to be used in the TRPC server codebase itself, for integrating testing... which means `commons` would import `backend` code in order to get the `AppRouter` definition from the backend, while `backend` would import `commons` to get access to the factory.

was struggling until i found your answer.
